src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -58,6 +59,78 @@ namespace ErrorCodes
     extern const int HAVE_DEPENDENT_OBJECTS;
 }
 
+    class DatabaseNameHints : public IHints<1, DatabaseNameHints>
+    {
+    public:
+        DatabaseNameHints(const DatabaseCatalog & database_catalog_)
+                : database_catalog(database_catalog_)
+        {
+        }
+    private:
+        Names getAllRegisteredNames() const override;
+        const DatabaseCatalog & database_catalog;
+    };
+
+    class TableNameHints : public IHints<1, TableNameHints>
+    {
+    public:
+        TableNameHints(const DatabaseCatalog & database_catalog_, ContextPtr context_, const String & database_name_)
+                : database_catalog(database_catalog_)
+                , context(context_)
+                , database_name(database_name_)
+        {
+        }
+    private:
+        Names getAllRegisteredNames() const override;
+        const DatabaseCatalog & database_catalog;
+        ContextPtr context;
+        String database_name;
+    };
+
+    Names DatabaseNameHints::getAllRegisteredNames() const
+    {
+        Names result;
+        auto database_instances = database_catalog.getDatabases();
+        for (const auto & database_name : database_instances | boost::adaptors::map_keys)
+        {
+            if (database_name == DatabaseCatalog::TEMPORARY_DATABASE)
+                continue;
+            result.emplace_back(database_name);
+        }
+        return result;
+    }
+
+    Names TableNameHints::getAllRegisteredNames() const
+    {
+        Names result;
+        DatabasePtr database = database_catalog.tryGetDatabase(database_name);
+        if (database)
+        {
+            for (auto table_it = database->getTablesIterator(context); table_it->isValid(); table_it->next())
+            {
+                const auto & storage_id = table_it->table()->getStorageID();
+                result.emplace_back(storage_id.getTableName());
+            }
+        }
+        return result;
+    }
```

Comment:
Please add a reasonable limit to avoid complexity in case of many created tables or databases (e.g. 1000).